jobserver/github.py

```diff
@@ -102,3 +104,46 @@ def get_repos_with_branches():
             "url": repo["url"],
             "branches": branches,
         }
+
+
+class GithubOrganizationOAuth2(GithubOAuth2):
+    """Github OAuth2 authentication backend for organizations"""
+
+    name = "github-org"
+    no_member_string = "User doesn't belong to the organization"
+
+    def user_data(self, access_token, *args, **kwargs):
+        """
+        Check the User is part of our Org
+
+        This is a near-complete reimplementation of GithubMemberOAuth2's
+        .user_data() which gets the correct URL from GithubOrganizationOAuth2's
+        .member_url().
+
+        We use our service-level PAT to avoid requesting further scopes from
+        the user (specifically admin:org) to make the call to the Org
+        membership endpoint.
+        """
+        user_data = super().user_data(access_token, *args, **kwargs)
+
+        f = furl(BASE_URL)
+        f.path.segments += [
+            "orgs",
+            "opensafely",
+            "members",
+            user_data.get("login"),
+        ]
+
+        # Use our "service-level" PAT instead of the User's OAuth token to
+        # avoid asking for the extra admin:org scope just for this check
+        headers = {"Authorization": f"token {TOKEN}"}
+
+        try:
+            self.request(f.url, headers=headers)
+        except requests.HTTPError as err:
+            # if the user is a member of the organization, response code
+            # will be 204, see http://bit.ly/ZS6vFl
```

Comment:
I think this should be a link to https://docs.github.com/en/free-pro-team@latest/rest/reference/orgs#check-organization-membership-for-a-user ?

I'd find the link & comment (perhaps with "Github API call to check if a user is, publicly or privately, a member of the organization") more useful a few lines above next to `f.path.segments`; when reading the code I was like "what's this thing mean"?!

Reply:
I probably should have marked this section in a more obvious way!  This method was taken verbatim taken from social-core's `GithubMemberOAuth2` class with only the URL and token parts changed.  I seriously considered entirely rewriting the whole thing (there's a branch which can never happen in the exception handler!) but opted for miniminal changes from their code to support fixing "wtf is going on here" later down the line.

Reply:
This comment is specifically calling out that conditional below which checks for non-204 responses, I'm not sure why they chose to link the docs there instead of the two different subclasses which construct the URL!

Reply:
Sorry, I think there's something I'm not understanding...!

For me, at least, that bit.ly URL resolves to https://docs.github.com/en/free-pro-team@latest/rest/reference/orgs#members, which is not the endpoint being called; the URL I suggested above _is_ for the endpoint being called and does describe the various status codes that can be returned?

Reply:
Ah, yes, the new GitHub docs deep linking is a bit trash.  It's linked to the Members section, not any particular endpoint but the offset dumps you part way into the first endpoint 🤪 

On pondering this more last night I realised keeping it tied to the social-core implementation without more specific comments is going to (and already has as evidenced by this thread!) cause more confusion so I'll rewrite this to be clearer and link to the specific endpoint!